File: stacks/dynamic-dashboard/config-driven-shiny/app/catalog/schema.py
"""The PlotSpec, in Pydantic — the same contract the nextjs stack states in Zod.

Every field is a decision the config author may make; every absent field is one they
may not. The PlotSpec models carry no free text. The *Card* models wrap a PlotSpec with
`kind` (the tool name, explicit because there is no tool call) and an optional human
`title` — copy a person writes is allowed, copy a model writes was not.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Discriminator, Field, Tag, TypeAdapter, create_model

from .datasets import DATASETS

FILTER_OPS = ("==", "!=", ">", ">=", "<", "<=", "in")
FilterOp = Literal[FILTER_OPS]  # type: ignore[valid-type]


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class CountMeasure(_Strict):
    op: Literal["count"]


def _spec_models(dataset: str) -> tuple[type[BaseModel], type[BaseModel]]:
    """Build the scatter and bar PlotSpec models for one dataset from its column catalog."""
    d = DATASETS[dataset]
    numeric = Literal[d.numeric]  # type: ignore[valid-type]
    categorical = Literal[d.categorical]  # type: ignore[valid-type]
    any_col = Literal[d.columns]  # type: ignore[valid-type]

    filter_model = create_model(
        f"{dataset.title()}Filter",
        __base__=_Strict,
        column=(any_col, ...),
        op=(FilterOp, Field(description="'in' takes a list value; the rest take a scalar")),
        value=(int | float | str | list[str], Field(..., description="scalar, or a list of 1–10 strings for 'in'")),
    )
    agg_model = create_model(
        f"{dataset.title()}AggMeasure",
        __base__=_Strict,
        op=(Literal["mean", "median"], ...),
        column=(numeric, ...),
    )
    measure = Annotated[CountMeasure | agg_model, Field(discriminator="op")]
    filters = (list[filter_model] | None, Field(None, max_length=3))  # type: ignore[valid-type]

    scatter = create_model(
        f"{dataset.title()}Scatter",
        __base__=_Strict,
        dataset=(Literal[dataset], Field(..., description=d.label)),
        x=(numeric, ...),
        y=(numeric, ...),
        color=(categorical | None, Field(None, description="Colour points by this category")),
        filter=filters,
    )
    bar = create_model(
        f"{dataset.title()}Bar",
        __base__=_Strict,
        dataset=(Literal[dataset], Field(..., description=d.label)),
        by=(categorical, Field(..., description="One bar per level of this column")),
        measure=(measure, Field(..., description="Bar height: row count, or mean/median of a numeric column")),
        color=(categorical | None, Field(None, description="Split each bar by this category (grouped bars)")),
        filter=filters,
    )
    return scatter, bar


PenguinsScatter, PenguinsBar = _spec_models("penguins")
MtcarsScatter, MtcarsBar = _spec_models("mtcars")

ScatterSpec = Annotated[PenguinsScatter | MtcarsScatter, Field(discriminator="dataset")]
BarSpec = Annotated[PenguinsBar | MtcarsBar, Field(discriminator="dataset")]

ScatterSpecAdapter: TypeAdapter[Any] = TypeAdapter(ScatterSpec)
BarSpecAdapter: TypeAdapter[Any] = TypeAdapter(BarSpec)

# ── Cards: PlotSpec + kind + optional human title, flat in the YAML ──────────────────


def _card_model(kind: str, spec_model: type[BaseModel]) -> type[BaseModel]:
    return create_model(
        f"{spec_model.__name__}Card",
        __base__=spec_model,
        kind=(Literal[kind], ...),
        title=(str | None, Field(None, max_length=60, description="Optional human-written card title")),
    )


PenguinsScatterCard = _card_model("scatter", PenguinsScatter)
MtcarsScatterCard = _card_model("scatter", MtcarsScatter)
PenguinsBarCard = _card_model("bar", PenguinsBar)
MtcarsBarCard = _card_model("bar", MtcarsBar)


def _card_tag(v: Any) -> str | None:
    if isinstance(v, dict):
        return f"{v.get('kind')}:{v.get('dataset')}"
    return f"{getattr(v, 'kind', None)}:{getattr(v, 'dataset', None)}"


Card = Annotated[
    Annotated[PenguinsScatterCard, Tag("scatter:penguins")]
    | Annotated[MtcarsScatterCard, Tag("scatter:mtcars")]
    | Annotated[PenguinsBarCard, Tag("bar:penguins")]
    | Annotated[MtcarsBarCard, Tag("bar:mtcars")],
    Discriminator(_card_tag),
]
CardAdapter: TypeAdapter[Any] = TypeAdapter(Card)


def spec_of(card: BaseModel) -> dict[str, Any]:
    """The PlotSpec inside a card: what the LLM would have emitted, nothing more."""
    return card.model_dump(exclude={"kind", "title"}, exclude_none=True)


class DashboardHeader(_Strict):
    """The dashboard-level knobs. `cards` stays raw here so each card validates on its own."""

    title: str = Field("Dashboard", max_length=80)
    columns: Literal[1, 2] = 2
    cards: list[dict[str, Any]] = Field(default_factory=list)

File: stacks/dynamic-dashboard/genui-controlled-nextjs/src/catalog/schema.ts
/**
 * The PlotSpec — the catalog contract.
 *
 * These Zod schemas are the tool `inputSchema`s the model fills. Everything the model is
 * allowed to decide is a field here; everything it is not allowed to decide is absent.
 * The same schemas will be what a YAML file satisfies in the config-driven sibling.
 *
 * No free-text field: a first `just prove` run with an optional `title` showed it was the
 * only prop that ever drifted (see README → Evidence). Titles are derived from the spec.
 */
import { z } from "zod";
import { DATASETS } from "./datasets";

const FILTER_OPS = ["==", "!=", ">", ">=", "<", "<=", "in"] as const;
export type FilterOp = (typeof FILTER_OPS)[number];

type Cols = readonly [string, ...string[]];

function filterFor<const C extends Cols>(cols: C) {
  return z
    .object({
      column: z.enum(cols),
      op: z.enum(FILTER_OPS).describe("'in' takes a string[] value; the rest take a scalar"),
      value: z.union([z.number(), z.string(), z.array(z.string()).min(1).max(10)]),
    })
    .describe("One row filter; filters are ANDed");
}

function specsFor<const D extends string, const N extends Cols, const C extends Cols>(
  dataset: D,
  def: { label: string; numeric: N; categorical: C },
) {
  const all = [...def.numeric, ...def.categorical] as unknown as readonly [N[number] | C[number], ...(N[number] | C[number])[]];
  const filter = z.array(filterFor(all)).max(3).optional();

  const scatter = z.object({
    dataset: z.literal(dataset).describe(def.label),
    x: z.enum(def.numeric),
    y: z.enum(def.numeric),
    color: z.enum(def.categorical).optional().describe("Colour points by this category"),
    filter,
  });

  const bar = z.object({
    dataset: z.literal(dataset).describe(def.label),
    by: z.enum(def.categorical).describe("One bar per level of this column"),
    measure: z
      .discriminatedUnion("op", [
        z.object({ op: z.literal("count") }),
        z.object({ op: z.enum(["mean", "median"]), column: z.enum(def.numeric) }),
      ])
      .describe("Bar height: row count, or mean/median of a numeric column"),
    color: z.enum(def.categorical).optional().describe("Split each bar by this category (grouped bars)"),
    filter,
  });

  return { scatter, bar };
}

const penguins = specsFor("penguins", DATASETS.penguins);
const mtcars = specsFor("mtcars", DATASETS.mtcars);

export const ScatterSpec = z.discriminatedUnion("dataset", [penguins.scatter, mtcars.scatter]);
export const BarSpec = z.discriminatedUnion("dataset", [penguins.bar, mtcars.bar]);

export type ScatterSpec = z.infer<typeof ScatterSpec>;
export type BarSpec = z.infer<typeof BarSpec>;
export type Filter = { column: string; op: FilterOp; value: number | string | string[] };
export type PlotSpec = ScatterSpec | BarSpec;
